### DISCUSSION TITLE
Insert lines into CCD XML

### DISCUSSION THREAD

**Discussion by pbhuddleston** [nextgenhealthcare/connect#5224] [Q&A] [answered]
I need to add two lines to messages so they will process correctly. The original source is as folllows:

<ClinicalDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="urn:hl7-org:v3" xmlns:voc="urn:hl7-org:v3/voc" xmlns:sdtc="urn:hl7-org:sdtc">

 <realmCode code="US"/>
			  <typeId root="2.16.840.1.113883.1.3" extension="POCD_HD000040"/>
			  <templateId root="2.16.840.1.113883.10.20.22.1.1" extension="2015-08-01" />
			  <templateId root="2.16.840.1.113883.10.20.22.1.1"/>
			
			  <id extension="TT988" root="2.16.840.1.113883.19.5.99999.1" />

I need to add these last two lines listed here  before the "id extension=TT988" shown above so it looks like this...

<ClinicalDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="urn:hl7-org:v3" xmlns:voc="urn:hl7-org:v3/voc" xmlns:sdtc="urn:hl7-org:sdtc">

<?xml-stylesheet type="text/xsl" href="cda.xsl"?><ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:sdtc="urn:hl7-org:sdtc" xmlns:voc="urn:hl7-org:v3/voc" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <realmCode code="US"/>
    <typeId extension="POCD_HD000040" root="2.16.840.1.113883.1.3"/>
    <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.1"/>
    <templateId root="2.16.840.1.113883.10.20.22.1.1"/>
    <templateId root="2.16.840.1.113883.10.20.22.1.2"/>
    <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.2"/>

How do I do this??

**Comment by pacmano1** [comment]
Not a clear question.   You write "these last two lines" and show six lines.

Show a before and after please.

And please define "process correctly".   Are you receiving that file?  Sending it?  Process by who?

**Reply by pbhuddleston** [reply]
Sorry for the confusion, it seems my whole message only shows when I edit. I will include a text file with the entire message also

This is my original XML message from the clinic:

<?xml-stylesheet type="text/xsl" href="cda.xsl"?><ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:sdtc="urn:hl7-org:sdtc" xmlns:voc="urn:hl7-org:v3/voc" xmlns:xsi=http://www.w3.org/2001/XMLSchema-instance>

<realmCode code="US"/>
               <typeId root="2.16.840.1.113883.1.3" extension="POCD_HD000040"/>
                                <templateId root="2.16.840.1.113883.10.20.22.1.1" extension="2015-08-01" />
                              <templateId root="2.16.840.1.113883.10.20.22.1.1"/>
                                             
                              <id extension="TT988" root="2.16.840.1.113883.19.5.99999.1" />
                                               
I need to add these two lines before “id extension” to all messages:

                 <templateId root="2.16.840.1.113883.10.20.22.1.2"/>
                <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.2"/>

The completed message should look like this so it will process correctly:

<?xml-stylesheet type="text/xsl" href="cda.xsl"?><ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:sdtc="urn:hl7-org:sdtc" xmlns:voc="urn:hl7-org:v3/voc" xmlns:xsi=http://www.w3.org/2001/XMLSchema-instance>

               <realmCode code="US"/>
                             <typeId extension="POCD_HD000040" root="2.16.840.1.113883.1.3"/>
                              <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.1"/>
                              <templateId root="2.16.840.1.113883.10.20.22.1.1"/>
                              <templateId root="2.16.840.1.113883.10.20.22.1.2"/>
                              <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.2"/>
                              <id extension="TT988" root="2.16.840.1.113883.19.5.99999.1" />

Thank you for any help you can provide.

Pam H.
[Line issues.txt](https://github.com/nextgenhealthcare/connect/files/8826516/Line.issues.txt)

**Comment by pacmano1** [comment, marked as answer]
I promise I am not giving you are hard time.

Is your ask this:

I receive a CDA from an external source, I have not posted the entire contents of the CDA on this post.  In order to process that CDA in my channel I need to add some extra XML before the line:

`<id extension="TT988" root="2.16.840.1.113883.19.5.99999.1" />`

If that is your case, as simple way is to use the preprocessor to do a regex replace.

e.g. see https://regex101.com/r/unsr4a/1

There are examples on the forums on using a regex, I only shared a possible candidate for the search and replace.

**Comment by pbhuddleston** [comment]
Sorry for the confusion, I had originally put the 2 lines I need to add in red, guess that did not carry through.

This is my original XML message from the clinic:

<?xml-stylesheet type="text/xsl" href="cda.xsl"?><ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:sdtc="urn:hl7-org:sdtc" xmlns:voc="urn:hl7-org:v3/voc" xmlns:xsi=http://www.w3.org/2001/XMLSchema-instance>

<realmCode code="US"/>
               <typeId root="2.16.840.1.113883.1.3" extension="POCD_HD000040"/>
                                <templateId root="2.16.840.1.113883.10.20.22.1.1" extension="2015-08-01" />
                              <templateId root="2.16.840.1.113883.10.20.22.1.1"/>

                              <id extension="TT988" root="2.16.840.1.113883.19.5.99999.1" />

I need to add these two lines before “id extension” to all messages:

                 <templateId root="2.16.840.1.113883.10.20.22.1.2"/>
                <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.2"/>

The completed message should look like this so it will process correctly:

<?xml-stylesheet type="text/xsl" href="cda.xsl"?><ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:sdtc="urn:hl7-org:sdtc" xmlns:voc="urn:hl7-org:v3/voc" xmlns:xsi=http://www.w3.org/2001/XMLSchema-instance>

               <realmCode code="US"/>
                             <typeId extension="POCD_HD000040" root="2.16.840.1.113883.1.3"/>
                              <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.1"/>
                              <templateId root="2.16.840.1.113883.10.20.22.1.1"/>
                              <templateId root="2.16.840.1.113883.10.20.22.1.2"/>
                              <templateId extension="2015-08-01" root="2.16.840.1.113883.10.20.22.1.2"/>
                              <id extension="TT988" root="2.16.840.1.113883.19.5.99999.1" />

Thank you for any help you can provide.

Pam H.

**Reply by pacmano1** [reply]
That doesn't change how to do that - use a regex.  You will just modify the version I linked to at regex101 and call a .`replace()`
